### PR TITLE
New constraints design

### PR DESCRIPTION
**Pull request by MilosKozak**
Add feature to provide reason for every constraint
image

### CODE REVIEW

**Comment by stavlor** [comment]
:+1:

File: app/src/main/java/info/nightscout/androidaps/plugins/ConfigBuilder/ConfigBuilderPlugin.java [hunk 1]
```diff
@@ -451,16 +449,14 @@ public void applySMBRequest(APSResult request, Callback callback) {
...
-    public boolean isLoopEnabled() {
-        boolean result = true;
+    public void limitRunningLoop(BooleanConstraint value) {
 
         ArrayList<PluginBase> constraintsPlugins = MainApp.getSpecificPluginsListByInterface(ConstraintsInterface.class);
         for (PluginBase p : constraintsPlugins) {
             ConstraintsInterface constrain = (ConstraintsInterface) p;
             if (!p.isEnabled(PluginBase.CONSTRAINTS)) continue;
-            result = result && constrain.isLoopEnabled();
+            constrain.limitRunningLoop(value);
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/plugins/ConfigBuilder/ConfigBuilderPlugin.java]: Could fix the typo, constrain**t** here and in the package name? My mind trips over this every time ;-)

File: app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package info.nightscout.androidaps.interfaces.constrains;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by mike on 19.03.2018.
+ */
+
+public class Constraint {
```
Comment (jotomo) [inline app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java]: If you generify this, BooleanConstraint, IntegerConstraint (...) won't be needed :-)

File: app/src/main/java/info/nightscout/androidaps/plugins/ConfigBuilder/ConfigBuilderPlugin.java (hunk 1 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/plugins/ConfigBuilder/ConfigBuilderPlugin.java]: already done in next PR , thx anyway

File: app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java (hunk 2 above)
Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java]: that was my 1st version :) but this is somehow cleaner

Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java]: i'll think about it again

Comment (jotomo) [reply app/src/main/java/info/nightscout/androidaps/interfaces/constrains/Constraint.java]: ... or just refactor **if** it still makes sense when we actually need more types and see more concrete requirements for the type.

File: app/src/main/java/info/nightscout/androidaps/MainApp.java [hunk 3]
```diff
@@ -191,13 +194,14 @@ public void onCreate() {
...
         NSUpload.uploadAppStart();
+
         if (Config.NSCLIENT)
             FabricPrivacy.getInstance().logCustom(new CustomEvent("AppStart-NSClient"));
         else if (Config.G5UPLOADER)
             FabricPrivacy.getInstance().logCustom(new CustomEvent("AppStart-G5Uploader"));
         else if (Config.PUMPCONTROL)
             FabricPrivacy.getInstance().logCustom(new CustomEvent("AppStart-PumpControl"));
-        else if (MainApp.getConfigBuilder().isClosedModeEnabled())
+        else if (MainApp.getConstraintChecker().limitClosedLoop(new Constraint<>(true)).get())
```
Comment (AdrianLxM) [inline app/src/main/java/info/nightscout/androidaps/MainApp.java]: The wording is confusing here. "limitClosedLoop(...true)" sounds like the command to limit a closed loop.

Comment (MilosKozak) [reply app/src/main/java/info/nightscout/androidaps/MainApp.java]: what would you recommend?

**Comment by AdrianLxM** [comment]
I would call it "isClosedModeEnabled" as before.

The ConstraintChecker does not necessarily have to implement the same interface as all Plugins that do imposte constraints. this way the dummy `new Constraint<>(true)` could be called within the ContraintChecker.

Should I do this in code to make it more clear what I mean?

**Comment by MilosKozak** [comment]
and the interface function leave `limitClosedLoop`?

**Comment by AdrianLxM** [comment]
@MilosKozak this is what I meant: https://github.com/MilosKozak/AndroidAPS/commit/01a9e7682750908bb7adf5f315eb6e810088e312

**Comment by MilosKozak** [comment]
make PR

**Comment by AdrianLxM** [comment]
to dev or to this branch?

**Comment by MilosKozak** [comment]
this

**Pull request merged**